ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/LedgerDB/DbChangelog.hs

```diff
@@ -0,0 +1,427 @@
+    rollbacks <- chooseInt (0, AS.length (changelogVolatileStates dblog))
+    pure $ DbChangelogTestSetupWithRollbacks
+      { testSetup = setup
+      , rollbacks = rollbacks
+      }
+
+emptyDbChangelogAtSlot :: WithOrigin SlotNo -> DbChangelog TestLedger
+emptyDbChangelogAtSlot slotNo = emptyDbChangeLog (TestLedger ApplyEmptyMK $ pointAtSlot slotNo)
+
+resultingDbChangelog :: DbChangelogTestSetup -> DbChangelog TestLedger
+resultingDbChangelog setup = applyOperations (operations setup) (originalDbChangelog setup)
+
+applyOperations :: (TableStuff l, GetTip (l EmptyMK))
+  => [Operation l] -> DbChangelog l -> DbChangelog l
+applyOperations ops dblog = foldr' apply' dblog ops
+  where apply' (Extend newState) dblog' = extendDbChangelog dblog' newState
+        apply' (Prune sp) dblog'        = pruneVolatilePartDbChangelog sp dblog'
+
+
+{-------------------------------------------------------------------------------
+  Invariants
+-------------------------------------------------------------------------------}
+
+volatileTipAnchorsImmutable :: (GetTip (l EmptyMK), Eq (l EmptyMK)) => DbChangelog l -> Bool
+volatileTipAnchorsImmutable DbChangelog { changelogImmutableStates, changelogVolatileStates } =
+  AS.anchor changelogVolatileStates == AS.headAnchor changelogImmutableStates
+
+immutableAnchored :: DbChangelog TestLedger -> Bool
+immutableAnchored DbChangelog { changelogDiffAnchor, changelogImmutableStates } =
+  changelogDiffAnchor == fmap Point.blockPointSlot point
+  where point = getPoint $ getTip $ unDbChangelogState $ AS.anchor $ changelogImmutableStates
+
+sameNumberOfDiffsAsStates :: DbChangelog TestLedger -> Bool
+sameNumberOfDiffsAsStates dblog = AS.length imm + AS.length vol == lengthSeqUtxoDiff diffs
+  where imm = changelogImmutableStates dblog
+        vol = changelogVolatileStates dblog
+        ApplySeqDiffMK diffs = unTestTables $ changelogDiffs dblog
+
+checkInvariants :: DbChangelog TestLedger -> Bool
+checkInvariants dblog = volatileTipAnchorsImmutable dblog &&
+                        immutableAnchored dblog &&
+                        sameNumberOfDiffsAsStates dblog
+
+
+{-------------------------------------------------------------------------------
+  Properties
+-------------------------------------------------------------------------------}
+
+prop_emptySatisfiesInvariants :: Property
+prop_emptySatisfiesInvariants =
+  property $ checkInvariants (emptyDbChangelogAtSlot Origin)
+
+prop_generatedSatisfiesInvariants :: DbChangelogTestSetup -> Property
+prop_generatedSatisfiesInvariants setup =
+  property $ checkInvariants (resultingDbChangelog setup)
+
+prop_flushingKeepsImmutableTip :: DbChangelogTestSetup -> Property
+prop_flushingKeepsImmutableTip setup =
+  property $ (toKeepTip == toFlushTip) && (toFlushTip == dblogTip)
+  where
+    dblog = resultingDbChangelog setup
+    (toFlush, toKeep) = flushDbChangelog DbChangelogFlushAllImmutable dblog
+    dblogTip = youngestImmutableSlotDbChangelog dblog
```

Comment:
Aligning the `=`'s might make it clear that we're applying the same operation in all three cases.